sage.py

```diff
@@ -481,483 +449,311 @@ def getepisodes(action_seq, mcat, plot, debug=False):
-                            rest = (end+1, len(ep)-1)
-                            al = episodes[start:end+1]
-                            if len(al) < 2:
-                                #print('discarding symbol ', [x[2] for x in al])#, start, end, len(episodes))
-                                continue
-                            att = atta+'->'+victim+'-'+str(c)
-                            #print('---', att, [x[2] for x in al])
-                            keys.append(att)
-                            alerts.append(al)
-                            c+=1
-                        #print('--', ep[rest[0]: rest[1]+1])
-                        al = episodes[rest[0]: rest[1]+1]
-                        if len(al) < 2:
-                            #print('discarding symbol ', [x[2] for x in al]) # TODO This one is not cool1
-                            continue
-                        att = atta+'->'+victim+'-'+str(c)
-                        #print('---', att, [x[2] for x in al])
-                        keys.append(att)
-                        alerts.append(al)
-    print('\n# sub-sequences', len(keys))
-    return (alerts, keys)
-
-## 27 Aug 2020: Generating traces for flexfringe 
-def generate_traces(alerts, keys, datafile, test_ratio=0.0):
-    victims = alerts
-    al_services = [[most_frequent(y[6]) for y in x] for x in victims]
-    print('----- All unique servcies -----')
-    print(set([item for sublist in al_services for item in sublist]))
+            # Cut episode sequence when a low-severity episode follows a high-severity episode
+            count = 0
+            mcats = [epi[2] for epi in episodes]
+            cuts = [i for i in range(len(episodes) - 1) if (len(str(mcats[i])) > len(str(mcats[i + 1])))]  # (ep[i] > 100 and ep[i+1] < 10)]
+
+            rest = (0, len(episodes) - 1)
+            for j in range(len(cuts)):
+                start = 0 if j == 0 else cuts[j - 1] + 1
+                end = cuts[j]
+                rest = (end + 1, len(episodes) - 1)
+                subsequence = episodes[start:end+1]
+                if len(subsequence) < 2:
+                    continue
+                _subsequences[attacker_victim + '-' + str(count)] = subsequence
+                count += 1
+            subsequence = episodes[rest[0]:rest[1]+1]
+            if len(subsequence) < 2:
+                # print('discarding symbol ', [x[2] for x in al]) # TODO This one is not cool1
+                continue
+            _subsequences[attacker_victim + '-' + str(count)] = subsequence
+
+    print('\n# sub-sequences:', len(_subsequences))
+    return _subsequences
+
+
+# Step 4.2: Generate traces for FlexFringe (27 Aug 2020)
+def generate_traces(subsequences, datafile):
+    all_services = [[_most_frequent(epi[6]) for epi in subseq] for subseq in subsequences.values()]
+    print('----- All unique services -----')
+    print(set([service for services_subseq in all_services for service in services_subseq]))
```

Comment:
What exactly is the purpose of this part? To compute all services present in the episodes or only the ones present in the AGs (i.e. mserv's)? In the former case, I don't get the use of `most_frequent`. In the latter case, the log is a bit misleading, also this can be computed in the for-loop below where mserv's are actually computed